tket/src/Mapping/MappingFrontier.cpp

```diff
@@ -596,4 +586,54 @@ void MappingFrontier::merge_ancilla(
   this->bimaps_->final.left.erase(merge);
 }
 
+bool MappingFrontier::valid_boundary_operation(
+    const ArchitecturePtr& architecture, const Op_ptr& op,
+    const std::vector<Node>& uids) const {
+  // boxes are never allowed
+  OpType ot = op->get_type();
+  if (is_box_type(ot)) {
+    return false;
+  }
+
+  if (ot == OpType::Conditional) {
+    OpType cond_ot = static_cast<const Conditional&>(*op).get_op()->get_type();
+    // conditional boxes are never allowed, too
```

Comment:
What about conditional conditional boxes ... ? Do we need recursion here?

Reply:
added recrusion, scary